### PR TITLE
Take into account units of `initial` argument to reductions

### PR DESCRIPTION
**Pull request by mhvk**
This is needed for `np.add.reduce` (and thus `np.sum`) and others.

This partially addresses the numerous failures with `numpy-dev`, but while those are the reason the bug was found, the bug is present for all numpy versions (most new tests under `test_sum_initial` fail on current master). Hence, the `numpy-dev` job should be ignored for this PR.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/units/quantity.py [hunk 1]
```diff
@@ -1862,7 +1894,11 @@ def nansum(self, axis=None, out=None, keepdims=False):
             return self._wrap_function(np.nansum, axis,
                                        out=out, keepdims=keepdims)
     else:
+        # TODO: deprecate this method? It is not on ndarray, and we do not
```
Comment (pllim) [inline astropy/units/quantity.py:1897]: Is there a follow-up issue for this?

**Comment by seberg** [comment]
Ah, so this is due to me tweaking `arr.fill()`, which gets used for `initial=` internally.  Just to make sure:  This is OK to keep in for the next NumPy release, or should we look into a "fix" in NumPy?  In the end, I guess without this fix it fails even when the units match?

**Comment by pllim** [comment]
There are still 29 errors. Is there a plan to handle those? I am not sure which ones will go away when Numpy backs out the backport and which one is the other one where maybe we can just ignore the warning for.

**Comment by mhvk** [comment]
@seberg - I think here your update actually uncovered a bug, so no need to change that!

@pllim - I think we should judge this PR independently of the other numpy-dev failures, since it fixes a bug for all versions of numpy:
```
# without this
In [2]: np.sum([1, 2]*u.m, initial=1*u.cm)
Out[2]: <Quantity 4. m>

# with this
In [2]: np.sum([1, 2]*u.m, initial=1*u.cm)
Out[2]: <Quantity 3.01 m>
```

**Comment by seberg** [comment]
There is one more failure though in: `test_models_evaluate_with_units` which happens for the `Const?D` models.
The traceback is pretty confusing, but the core issue must be they use `x.fill(amplitude.item())` and that has modified logic as well:
https://github.com/astropy/astropy/blob/3eb38adc6698973468ec31278e6e21934e570081/astropy/modeling/functional_models.py#L1804

**Comment by pllim** [comment]
I was gonna approve but I should wait until @seberg 's concern is resolved.

**Comment by seberg** [comment]
Sorry, that is no concern!  It is just one more failure even after this and gh-13332 is resolved.
I suppose it might mean a fixup in NumPy, but since this is a bug-fix independent of NumPy that doens't even matter.

**Comment by mhvk** [comment]
p.s. Really, this is because `initial` is relatively new, and I hadn't appreciated that it would affect anything for `Quantity`. Even though I introduced `where` for the reductions in numpy and wrote the corresponding quantity tests...

File: astropy/units/quantity.py (hunk 1 above)
Comment (mhvk) [reply astropy/units/quantity.py:1897]: No, it is not urgent and has been around forever. So either I do it very soon or just whenever someone encounters this.......

**Comment by mhvk** [comment]
@pllim - just to be sure, this particular PR is really about units on `initial`, so best treated separately.

But, @seberg, thanks for pointing to another instance where a related problem bites!!

**Review by pllim — APPROVED** [review]
Okay, thanks for the fix and clarifications! Let's merge so we can reduce the number of errors and follow-up with more PRs.

**Comment by pllim** [comment]
@mhvk , are you able to do a manual backport to v5.0.x ? 🙏

**Pull request merged**